libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp

```diff
@@ -0,0 +1,191 @@
+}
+
+#define REG_CONFIGURATION 0x00
+#define REG_MANUFACTURER_ID 0xFE
+#define REG_DIE_ID 0xFF
+bool AP_BattMonitor_INA3221::init_at_address(uint8_t init_address)
+{
+    _dev = std::move(hal.i2c_mgr->get_device(_params._i2c_bus, init_address, 100000, true, 20));
+    if (!_dev) {
+        return false;
+    }
+
+    debug("INA3221: probe @0x%02x on bus %u", init_address, _params._i2c_bus.get());
+
+    WITH_SEMAPHORE(_dev->get_semaphore());
+
+    // check manufacturer_id
+    uint16_t manufacturer_id;
+    if (!read_register(REG_MANUFACTURER_ID, manufacturer_id)) {
+        debug("read register (%u (0x%02x)) failed", REG_MANUFACTURER_ID, REG_MANUFACTURER_ID);
+        return false;
+    }
+    if (manufacturer_id != 0x5449) {  // 8.6.1 p24
+        debug("Bad manufacturer_id: 0x%02x", manufacturer_id);
+        return false;
+    }
+
+    uint16_t die_id;
+    if (!read_register(REG_DIE_ID, die_id)) {
+        debug("Bad die: 0x%02x", manufacturer_id);
+        return false;
+    }
+    if (die_id != 0x3220) {  // 8.6.1 p24
+        return false;
+    }
+
+    debug("Found INA3221 @0x%02x on bus %u", init_address, _params._i2c_bus.get());
+
+    // reset the device:
+    union {
+        struct PACKED {
+            uint16_t mode : 3;
+            uint16_t shunt_voltage_conversiontime : 3;
+            uint16_t bus_voltage_conversiontime : 3;
+            uint16_t averaging_mode : 3;
+            uint16_t ch1_enable : 1;
+            uint16_t ch2_enable : 1;
+            uint16_t ch3_enable : 1;
+            uint16_t reset : 1;
+        } bits;
+        uint16_t word;
+    } configuration {
+            0b111, // continuous operation
+            0b111, // 8ms conversion time for shunt voltage
+            0b111, // 8ms conversion time for bus voltage
+            0b111, // 1024 samples / average
+            0b1,  // enable ch1
+            0b1,  // enable ch2
+            0b1,  // enable ch3
+            0b0    // don't reset...
+                };
+
+    if (!write_register(REG_CONFIGURATION, configuration.word)) {
+        return false;
+    }
+
+    _dev->register_periodic_callback(
+        100000,
+        FUNCTOR_BIND_MEMBER(&AP_BattMonitor_INA3221::_timer, void));
+
+    return true;
+}
+
+void AP_BattMonitor_INA3221::_timer(void)
+{
+    uint8_t reg_shunt;
+    uint8_t reg_bus;
+    switch (_type) {
+    case AP_BattMonitor::Type::INA3221_CH1:
+        reg_shunt = 1;
+        reg_bus = 2;
+        break;
+    case AP_BattMonitor::Type::INA3221_CH2:
+        reg_shunt = 3;
+        reg_bus = 4;
+        break;
+    case AP_BattMonitor::Type::INA3221_CH3:
+        reg_shunt = 5;
+        reg_bus = 6;
+        break;
+    default:
+        return;
+    }
+    uint16_t shunt_voltage;
+    if (!read_register(reg_shunt, shunt_voltage)) {
+        return;
+    }
+    uint16_t bus_voltage;
+    if (!read_register(reg_bus, bus_voltage)) {
+        return;
+    }
+
+    //transfer readings to front end:
+    WITH_SEMAPHORE(_sem);
```

Comment:
So you aren't wrong that there should be a semaphore around writing to the front end, as the SMBus drivers have this bug, but as far as I can tell you have gained no protection here, as you only grab this in your init, and in the timer. It doesn't actually get locked by the frontend ever when reading state out of the battery, so I think as it stands this is wasting processing time/memory. If you want to use this, you need to accumulate the results in your own internal copy here, and then take this semaphore in the `read()` call which currently has your debug print.

Reply:
Done